Rostest

joshuamorton opened the pull request:
Please tell me what to explain and comment on.

Fixes #113

joshuamorton committed:
Add utilities to make writing ros tests less painful.

joshuamorton committed:
Update test_inputer to use test utils.

joshuamorton committed:
Add some tests to the bearing node

joshuamorton committed:
Merge branch 'process-split' into rostest

Conflicts:
	buzzmobile/process/bearing/bearing.py

joshuamorton committed:
Added basic tests for bearing

joshuamorton committed:
Merge branch 'master' of github.com:gtagency/buzzmobile into rostest

joshuamorton committed:
Add required dependency for roslaunch

joshuamorton committed:
Add a roslaunch file that just inits params, for testing

joshuamorton committed:
Update rostest utils to handle cleanup

joshuamorton committed:
Update tests for new rostest util format

joshuamorton committed:
Make bearing use global rosparams

This fixes a testing issue.

joshuamorton committed:
Fix install for CI

irapha commented on buzzmobile/process/bearing/bearing.py:
```diff
@@ -24,13 +24,12 @@ def median(self):
 bearing_pub = rospy.Publisher('bearing', Float64, queue_size=1)
 
-MIN_FIX_DISTANCE = rospy.get_param('min_fix_distance')
+MIN_FIX_DISTANCE = rospy.get_param('/buzzmobile/min_fix_distance')
```
just a note that these changes here mean that we'll always have to load params using the new `params.launch`. It also means we have to change wherever we use params to have the `/buzzmobile/` prefix

joshuamorton replied:
You shouldn't need to run everything with params.launch (you just need to have the params loaded, which our other launch files do anyway).

The other alternative is to set defaults with `rospy.get_param('location', DEFAULT)`, which does the same thing more or less.

irapha replied:
So the thing is that our launch files are not namespacing the rosparam launch. Which is why everywhere we call get_param without '/buzzmobile/' prefix. I know we probably should think about namespacing everything, just like we're doing with nodes. But I think that should be in a separate PR. Can you make your launch file not namespace it, and revert this change in `get_param`? (also maybe make an issue for namespacing rosparams)

joshuamorton replied:
They do currently namespace everything, but put all running nodes in the same namespace. You can still use the explicit form. Its just that since I'm spinning up nodes outside of the namespace, I _must_ use the explicit form. That said, i think I can fix this elsewhere.

irapha replied:
Okay up to you. I just want to make sure this works when not in testing

joshuamorton replied:
Fixed, although its fragile.

irapha commented on buzzmobile/setup.py:
```diff
@@ -22,5 +22,6 @@
         'rospkg==1.0.41',
         'ds4drv', # requires apt-get install python-dev
         'empy==3.3.2', # required to make catkin_make work
+        'netifaces==0.10.5',
```
Is this needed? I don't see it being used anywhere

joshuamorton replied:
Testing wouldn't run without it, its used by like some dark part of rospy.

irapha replied:
............ can you add `# required to make rospy work`?

irapha commented on buzzmobile/tests/__init__.py:
```diff
@@ -0,0 +1 @@
+import tests.rostest_utils
```
have you checked if this is needed? I don't particularly care, and it won't block this PR but I'd like to know

joshuamorton replied:
Fixed.

irapha commented on buzzmobile/tests/rostest_utils.py:
```diff
@@ -0,0 +1,112 @@
+            launch = ROSLauncher(f)
+            launch.start()
+
+            # can we do something async here?
```
Wanna turn this into a `TODO(joshua)`? :)

joshuamorton replied:
So actually this is blocking, I'll just remove it because it doesn't matter.

irapha commented on buzzmobile/tests/sense/test_inputer.py:
```diff
@@ -1,55 +1,21 @@
-        self.run_node_with_callback(cb)
-        assert self.success
 
+    @rostest_utils.launch_node('buzzmobile', 'inputer.py', 'params.launch')
```
Is it a good idea to have `launch_node` manage starting launch files too? In my head we'd have a decorator for each. But I can see that maybe things won't be super guaranteed to run in order so idk.

irapha replied:
Alternatively, we could have something like `with_launch`, just like `with_roscore`

joshuamorton replied:
I think, but I'm not sure, that the launcher api requires that the launch files be tied to the node startup, I'm not actually sure why, but see the source for `launch_node`, it creates a Launcher tied to the launch file which is then used to launch the node. Conceivably we could create a 'global' launcher for multiple nodes, and that might be better, (change the api to be list of tuples of proj/node and then launch=launchfile as a kwarg) and have it spin all of them up. Would that be preferable?

irapha replied:
Oh that's interesting. What you're doing in the background is adding the node to the launch file and launching everything together, right?

In answer to your question, I think that might be preferable. My main concern in separating things is if we want to launch multiple nodes for an integration test, there is currently no easy way of both using a launch file (say, for rosparams), and starting up multiple nodes.

So either way would work as an api:

``` python
@launch_nodes(('buzzmobile', 'gps_mapper'), ('buzzmobile', 'lidar_to_frame'), ('buzzmobile', 'frame_merger'), launch='params.launch')
def test_merging_frames_integration(self):
    def cb(world_model):
        self.success = world_model is correct
    with check_topic('world_model', Image, cb):
        sleep(1) # or something, it'll take time to merge. not sure what's best practice
        assert self.success
```

or

``` python
@with_launch_file('params.launch')
@launch_node('buzzmobile', 'gps_mapper')
@launch_node('buzzmobile', 'lidar_to_frame')
@launch_node('buzzmobile', 'frame_merger')
def test_merging_frames_integration(self):
    def cb(world_model):
        self.success = world_model is correct
    with check_topic('world_model', Image, cb):
        sleep(1) # or something, it'll take time to merge. not sure what's best practice
        assert self.success
```

Note that in the second one, `@with_launch_file` is also optional. So running nodes without a launch file is possible

joshuamorton replied:
Fixed.

irapha commented on buzzmobile/tests/sense/test_inputer.py:
```diff
@@ -1,55 +1,21 @@
-if __name__ == '__main__':
-    rostest.rosrun(PKG, NAME, TestInputer)
-
+        with rostest_utils.test_node('direction', String, cb):
```
So `test_node` is way for us to check on a message that is maybe being published by the node launched in `launch_node`, correct? And if `launch_node` depends on anything, we can use a `mock_node`.

I know `test_node` creates a node that subscribes to `launch_node`'s message in the background, but it's kind of confusing to me to see a test that has:

``` python
test_node('message_name')
# instead of...
test_node('node_name')
```

What do you think of renaming `test_node` to something like `check_topic`, or `check_message`, etc... I don't really know what the best name would be..

irapha replied:
The other thing I'm thinking about is: suppose `launch_node` takes 5 sec to publish 'direction'. Won't the `with` statement run the `assert` immediately? Maybe that's not a problem, but maybe it'd be nice to have something like "wait for callback, with timeout". Thought I guess that can be done here and not in a utils method..

joshuamorton replied:
That's one of the biggest issue, I think that it might be better to refactor test node and mock node to just be topics, not nodes, and use `rospy.wait_for_message` which is a blocking call to mock the receiving node.

joshuamorton replied:
wait_for_message doesn't solve this. I think it can be solved by spinning up a temp background node to subscribe, receive, and be done, but that's really ugly.

irapha commented on buzzmobile/tests/rostest_utils.py:
```diff
@@ -0,0 +1,112 @@
+        time.sleep(1)
+
+@contextlib.contextmanager
+def mock_node(topic, rosmsg_type, queue_size=1):
```
I know this creates a mock node, but when I'm writing the test, I only really care that it will publish a topic, right? Why not rename it to `mock_topic` or something? (see other comment about `test_node`)

joshuamorton replied:
This is a good question, I may actually refactor all of this to just publish to a topic, although I'm not sure.

irapha replied:
k lmk

irapha replied:
lmk status on changing this. I made a comment somwhere else about it too

joshuamorton replied:
nevermind, the 'publish to a topic' just implicitly spins up a node and so gives us less control.

irapha replied:
Sure but the naming is still confusing imo

irapha reviewed (changes requested):
Some thoughts. I'd like opinions

joshuamorton commented:
Right, the second one works better as an api, but isn't technically
feasible unless we make it

```
launcher = Launcher('params.launch')

@launcher.launch_node('buzzmobile', 'gps_mapper')
...
```

At least, I'd need to do dark evil things to even approach making the
second work otherwise.

On Sat, Nov 12, 2016 at 3:44 PM Raphael Gontijo Lopes <
notifications@github.com> wrote:

> ## _@iRapha_ commented on this pull request.
> 
> In buzzmobile/tests/sense/test_inputer.py
> https://github.com/gtagency/buzzmobile/pull/121:
> 
> > -    @rostest_utils.launch_node('buzzmobile', 'inputer.py', 'params.launch')
> 
> Oh that's interesting. What you're doing in the background is adding the
> node to the launch file and launching everything together, right?
> 
> In answer to your question, I think that might be preferable. My main
> concern in separating things is if we want to launch multiple nodes for an
> integration test, there is currently no easy way of both using a launch
> file (say, for rosparams), and starting up multiple nodes.
> 
> So either way would work as an api:
> 
> @launch_nodes(('buzzmobile', 'gps_mapper'), ('buzzmobile', 'lidar_to_frame'), ('buzzmobile', 'frame_merger'), launch='params.launch')def test_merging_frames_integration(self):
>     def cb(world_model):
>         self.success = world_model is correct
>     with check_topic('world_model', Image, cb):
>         sleep(1) # or something, it'll take time to merge. not sure what's best practice
>         assert self.success
> 
> or
> 
> @with_launch_file('params.launch')@launch_node('buzzmobile', 'gps_mapper')@launch_node('buzzmobile', 'lidar_to_frame')@launch_node('buzzmobile', 'frame_merger')def test_merging_frames_integration(self):
>     def cb(world_model):
>         self.success = world_model is correct
>     with check_topic('world_model', Image, cb):
>         sleep(1) # or something, it'll take time to merge. not sure what's best practice
>         assert self.success
> 
> Note that in the second one, @with_launch_file is also optional. So
> running nodes without a launch file is possible
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/gtagency/buzzmobile/pull/121, or mute the thread
> https://github.com/notifications/unsubscribe-auth/AEz4AJ8UcISF5xAOGjYA6yLA8CT_O377ks5q9iVEgaJpZM4KwTYQ
> .

joshuamorton committed:
Implement requested change to node launch api

joshuamorton committed:
Remove unnecessary import in init

joshuamorton committed:
Revert to namespaced nodes

joshuamorton committed:
Updated setup.py

irapha commented on buzzmobile/process/bearing/bearing.py:
```diff
@@ -55,7 +55,7 @@ def bearing(fix):
 def bearing_node():
     rospy.init_node('bearing', anonymous=True)
-    rospy.Subscriber('/fix', NavSatFix, bearing)
+    rospy.Subscriber('fix', NavSatFix, bearing)
```
this `fix` needs to be `/fix`

If we are running our nodes in the `buzzmobile` namespace, this subscribe call is equivalent to `Subscriber('/buzzmobile/fix/', ...`. However, third-party nodes like Nmea Navsat will not be in the buzzmobile namespace, so we need to make explicit `/fix`

irapha commented on buzzmobile/tests/process/test_bearing.py:
```diff
@@ -19,15 +19,16 @@ def test_distance(self):
-        with rostest_utils.mock_node('fix', NavSatFix, queue_size=None) as fix_node:
-            with rostest_utils.test_node('bearing', Float64, callback) as tn:
+        with rostest_utils.mock_node('/buzzmobile/fix', NavSatFix, queue_size=None) as fix_node:
+            with rostest_utils.test_node('/buzzmobile/bearing', Float64, callback):
```
Have you considered changing the names to `mock_topic` and `check_topic`?

irapha replied:
Also why are you mocking/testing these topics in the `buzzmobile` namespace? Bearing does not subscribe or publish anything in the buzzmobile namespace, as per file above.

irapha reviewed (changes requested):
Just a few comments, to add to the literal `#comments` you have, plus the old change requests I had made.

joshuamorton committed:
Update with better api for testing nodes

joshuamorton committed:
Rename and make function calls shorter

joshuamorton committed:
Merge branch 'master' of github.com:gtagency/buzzmobile into rostest

Conflicts:
	buzzmobile/process/bearing/calculate_directions.py
	buzzmobile/tests/sense/test_inputer.py
	install

joshuamorton committed:
Updated test to work with bugfix in another commit

irapha commented on buzzmobile/tests/process/test_bearing.py:
```diff
@@ -0,0 +1,37 @@
+        def callback(s, data):
+            s.result = data.data
+
+        with mock_node('/buzzmobile/fix', NavSatFix, queue_size=None) as fix_node:
```
To make things clearer, I suggest you rename this to `mock_pub` or `mock_publisher`. Because we're not really mocking an entire node (which may include multiple pub sub calls) and we're not really mocking just the topic (because of the `send` method).

irapha replied:
Also, for this test to work, I think you might need to change this to `'/fix`, and not namespace it. That's how we're launching third-party nodes in master right now: not namespacing them.

joshuamorton replied:
Done

irapha commented on buzzmobile/tests/process/test_bearing.py:
```diff
@@ -0,0 +1,37 @@
+                fix_node.send(NavSatFix(None, None, 39.029128, -111.838257, None, None, None))
+
+            # check the output from the node
+            assert ct != None
```
does this owrk being outside the `with`? I thought when it exited the `with`, it cleaned up with `__exit__`.

Also have we figured out a way to wait a little bit for `bearing` to finish publishing after it receives the two messages?

joshuamorton replied:
The __exit__ is something that I can write, the object named is still around and can be referenced, its just that the exit method cleans up the connection to the topic, so for example any new sent things wouldn't be noticed.

And no, unfortunately.

irapha replied:
Okay I think the second part of this is blocking right? Otherwise we can't actually test anything involving ros

joshuamorton replied:
Well we can, it just involves using sleeps that are factored out instead of  something more complex.

Currently tests are being run on ros.

irapha commented on buzzmobile/launch/params.launch:
```diff
@@ -0,0 +1,5 @@
+<launch>
+    <group ns="buzzmobile">
+        <rosparam command="load" file="$(find buzzmobile)/constants.yaml"/>
+    </group>
```
So the way our main launch files work is they currently have things like:

```
<third party nodes>
<buzzmobile namespace>
    <rosparam constants>
    <include our nodes>
</buzzmobile namespace>
```

I suggest you remove the buzzmobile namespace from this `params.launch`, put it in `buzzmobile/launch/includes/`, modify `main.launch` (and other main files under `buzzmobile/launch/temp` to be like:

```
<third party nodes>
<buzzmobile namespace>
    <include params>
    <include our nodes>
</buzzmobile namespace>
```

and finally create test-specific launch files (somewhere like `buzzmobile/tests/launch/parms_test.launch`), which define a namespace and include the params, like such:

```
<buzzmobile namespace>
    <include params>
</buzzmobile namespace>
```

This way we can be consistent all around, reuse as much as possible, and still be able to define test-specific launch files.

joshuamorton replied:
Done

irapha commented on buzzmobile/process/bearing/bearing.py:
```diff
@@ -46,7 +45,7 @@ def bearing(fix):
     """Initializes bearing node."""
     rospy.init_node('bearing', anonymous=True)
-    rospy.Subscriber('/fix', NavSatFix, bearing)
+    rospy.Subscriber('fix', NavSatFix, bearing)
```
Can you revert this to be `/fix`? No third-party nodes are namespaced, and `fix` is published by a third party node. Also see comment on subscribing to `fix` below.

joshuamorton replied:
Done

irapha commented on buzzmobile/tests/rostest_utils.py:
```diff
@@ -0,0 +1,161 @@
+    `which roslaunch`. 
+    """
+    def __init__(self, files):
+        super().__init__()
```
Is pylint not complaining about this? Dang, I think I may have not activated it correctly...

IIRC it should be complaining about 'not enough arguments given to super'

irapha replied:
Can you maybe check that pylint is working by adding a ` # TODO unassigned` somewhere?

joshuamorton replied:
Technically it shouldn't complain about that on python3, because I'm using the correct form of super for python3 (they made it less awful in the 2 -> 3 switch). So that might be it.

irapha replied:
Actually the reason why it's not complaining is because pylint is not being properly installed in CI. See https://336-53988135-gh.circle-artifacts.com/0/tmp/circle-artifacts.RwLj3tB/style.txt

irapha commented on buzzmobile/tests/rostest_utils.py:
```diff
@@ -0,0 +1,161 @@
+# TODO(joshuamorton@gatech.edu): swap this to a Map[int, launcher], where the
+# int is incrementally generated and stored within the calling decorator. This
+# will prevent the issue of overwriting or cross writing if it becomes a
+# danger.
```
IIRC this TODO is blocking, right?

joshuamorton replied:
Ehhh, its optional and I don't actually think its a pressing concern. Its only a problem if we were to do something like

    @with_launch_file
    def test():
        @with_launch_file
        def other_test
            ...

Which I don't see as likely to happen.

irapha replied:
If we have multiple tests, does pytest not parallelize things? (And therefore lead to problems?)

joshuamorton replied:
Not by default, afaik.

irapha replied:
Okay. We won't block anything on this then, we can work on it later on

irapha replied:
Maybe make an issue then?

irapha commented on buzzmobile/tests/rostest_utils.py:
```diff
@@ -0,0 +1,161 @@
+            process = launch.launch(node)
+            # Beware this is a bit of a hack, and will currently not work if we
+            # want to run more than 1 node with the same name.
+            while not any(nn.split('/')[-1].startswith(name.replace('.', '_')) 
```
ew

joshuamorton replied:
pls gib better way

irapha replied:
so when you say "run more than one node with the same name", to me it sounds like you can't run multiple tests that test the same node at the same time. While this is probably okay, in the future we may want to have a way of namespacing tests themselves, so they don't interfere with eachother. You can even use the test name as the node namespace.

joshuamorton replied:
The problem is a bit larger than that, if a test, for whatever reason, needs 2 camera nodes to run, that will break. (Imagine a future where we had 2 video streams).

irapha replied:
I'd argue that isn't a problem bc ros doens't allow nodes with the same name anyways, does it? we can have as many cameras as we want, but if they're named the same, ROS will complain. So I don't see how we need to handle this scenario in testing. The only scenario I see needing handling is multiple tests running at the same time

joshuamorton replied:
So, imagine I create a mock `uvc_cam` node that sends some video we bagged, and then I make another one.

The first will be `uvc_cam_8673po247` and the second will be `uvc_cam_9871236`, then the second will never actually start. Note that I'm not using the whole node name, but only the part that doesn't include the random numbers.

irapha commented on buzzmobile/tests/sense/test_inputer.py:
```diff
@@ -1,61 +1,20 @@
+            try:
+                ct.success
+            except AttributeError:
+                assert True
```
lol

joshuamorton replied:
:3

irapha replied:
Wait but don't we want to actually be testing something here?

joshuamorton replied:
This is about all we can do in this case.

joshuamorton committed:
Make requested launch file changes

joshuamorton committed:
Update tests to use new test param name

joshuamorton committed:
Swap fix to be outside of buzzmobile namespace

joshuamorton committed:
Change mock_node to mock_pub

joshuamorton committed:
Removed useless test

joshuamorton committed:
Change to a flag from flaky time.sleeps.

This makes a few assumptions, namely that you're only testing on the
first value returned, anything beyond that will require more powerful
stiff or some personalization.

Also yield from is there to await the coroutine that is being used, if
that isn't there, the sleep while loop will just block the thread
forever and the callback will never get the chance to run. This way it
doesn't.

irapha commented on buzzmobile/tests/rostest_utils.py:
```diff
@@ -142,13 +143,25 @@ class TestNode:
+    def wait_for_message(self):
+        while not self.received:
+            time.sleep(.1)
+        yield
```
should we have a timeout that can be set?

```python
def wait_for_message(self, timeout=None):
```

joshuamorton committed:
Added timeout

irapha commented on buzzmobile/tests/rostest_utils.py:
```diff
@@ -145,9 +145,13 @@ def __init__(self, topic, msg_type):
             time.sleep(.1)
+            elapsed += .1
+        if elapsed >= timeout:
+            raise TimeoutError('Test timed out')
```
How about 'Timed out waiting for message'? More descriptive of where it timed out.

joshuamorton committed:
Merge branch 'master' of github.com:gtagency/buzzmobile into rostest

joshuamorton committed:
Misc pylint fixes

joshuamorton committed:
Removed extraneous assert

joshuamorton committed:
Refactor to reduce file size

joshuamorton committed:
Changed api to be a bit nicer

joshuamorton committed:
Remove unnecessary roscore invocation

joshuamorton committed:
Relatively unrelated bugfix of a pylint error

joshuamorton committed:
Update pylint script again

joshuamorton committed:
Another pylint fix attempt

joshuamorton committed:
Added import to __init__.py in testing module

joshuamorton committed:
Add default callback function to mock nodes

joshuamorton committed:
Some final pylint changes
